File: leetcode/cpp/24.swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(!head) return head;
        
        //反转后的表头肯定时第二个元素
        ListNode* res = head->next;
        
        //两两反转
        ListNode *curr = head->next;
        ListNode *prev = head;
        
        //需要前一个节点的指针将反转后的链表与前面的链表连接起来
        ListNode* conn = nullptr;
        //当两个节点都存在时
        while( curr && prev )
        {
            ListNode *t = curr->next;
            curr->next = prev;
            prev->next = t;
            if(conn){
                conn->next = curr;
                conn = prev;
            }
            else{
                conn = prev;
            }
            prev = t;
            
            if(prev)
                curr = t->next;
            
        }
        
        return res ? res : head;
    }
};

File: leetcode/python/24.swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        dummy = ListNode(-1)
        dummy.next = head

        prev = dummy

        while prev and prev.next and prev.next.next:
            curr = prev.next
            back = curr.next.next

           # 断开链表
            curr.next.next = None
            prev.next = self.reverseList(curr)
            curr.next = back

            prev = curr
        return dummy.next

    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        while head:
            next = head.next
            head.next = prev

            prev = head
            head = next

        return prev
